dotnet/src/Extensions/PromptTemplate.Handlebars/Helpers/SystemHelpers/KernelSystemHelpers.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.Json;
+using HandlebarsDotNet;
+using HandlebarsDotNet.Compiler;
+using HandlebarsDotNet.Helpers;
+
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+namespace Microsoft.SemanticKernel.PromptTemplates.Handlebars.Helpers;
+#pragma warning restore IDE0130
+
+/// <summary>
+/// Extension class to register additional helpers as Kernel System helpers.
+/// </summary>
+public static class KernelSystemHelpers
+{
+    /// <summary>
+    /// Register all (default) or specific categories.
+    /// </summary>
+    /// <param name="handlebarsInstance">The <see cref="IHandlebars"/>-instance.</param>
+    /// <param name="variables">Dictionary of variables maintained by the Handlebars context.</param>
+    /// <param name="options">Handlebars promnpt template options.</param>
+    public static void Register(IHandlebars handlebarsInstance, KernelArguments variables, HandlebarsPromptTemplateOptions options)
+    {
+        RegisterHandlebarsDotNetHelpers(handlebarsInstance, options);
+        RegisterSystemHelpers(handlebarsInstance, variables);
+    }
+
+    private static void RegisterHandlebarsDotNetHelpers(IHandlebars handlebarsInstance, HandlebarsPromptTemplateOptions helperOptions)
+    {
+        HandlebarsHelpers.Register(handlebarsInstance, optionsCallback: options =>
+        {
+            options.PrefixSeparator = helperOptions.PrefixSeparator;
+            options.Categories = helperOptions.Categories;
+            options.UseCategoryPrefix = helperOptions.UseCategoryPrefix;
+            options.CustomHelperPaths = helperOptions.CustomHelperPaths;
+        });
+    }
+
+    /// <summary>
+    /// Register all system helpers.
+    /// </summary>
+    /// <param name="handlebarsInstance">The <see cref="IHandlebars"/>-instance.</param>
+    /// <param name="variables">Dictionary of variables maintained by the Handlebars context.</param>
+    /// <exception cref="KernelException">Exception thrown when a message does not contain a defining role.</exception>
+    private static void RegisterSystemHelpers(
+        IHandlebars handlebarsInstance,
+        KernelArguments variables)
+    {
+        // TODO [@teresaqhoang]: Issue #3947 Isolate Handlebars Kernel System helpers in their own class
+        // Should also consider standardizing the naming conventions for these helpers, i.e., 'Message' instead of 'message'
+        handlebarsInstance.RegisterHelper("message", (writer, options, context, arguments) =>
+        {
+            var parameters = arguments[0] as IDictionary<string, object>;
+
+            // Verify that the message has a role
+            if (!parameters!.ContainsKey("role"))
```

Comment:
Casting with `as` followed by immediate usage is an anti-pattern.  Better to instead use a normal cast, e.g.
```C#
var parameters = (IDictionary<string, object>)arguments[0];

if (!parameters!.ContainsKey("role"))
```
If you use `as` and the type isn't of the right type, it'll result in the failure being a NullReferenceException. If you use a normal cast and the type isn't of the right type, it'll result in a cast exception that includes more helpful information.

Reply:
Ohhh good to know, changed it. Thank you!